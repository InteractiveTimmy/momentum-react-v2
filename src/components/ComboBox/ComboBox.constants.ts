const CLASS_PREFIX = 'md-combo-box';

const DEFAULTS = {};

const STYLE = {
  wrapper: `${CLASS_PREFIX}-wrapper`,
};

export { CLASS_PREFIX, DEFAULTS, STYLE };
