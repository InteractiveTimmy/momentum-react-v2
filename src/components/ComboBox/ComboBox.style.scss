.md-combo-box-wrapper {
  position: relative;
  display: flex;
  align-items: center;

  button {
    border-radius: 0;
  }

  .md-select-overlay {
    width: 100%;
    position: absolute;
    padding: 0;
    margin: 0;
    z-index: 9999;
    top: 3rem;
  }
}
