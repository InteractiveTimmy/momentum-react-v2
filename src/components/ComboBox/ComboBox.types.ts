import { CSSProperties } from 'react';
import type { ComboBoxProps } from '@react-types/combobox';

export interface Props<T> extends ComboBoxProps<T> {
  /**
   * Custom class for overriding this component's CSS.
   */
  className?: string;

  /**
   * Custom id for overriding this component's CSS.
   */
  id?: string;

  /**
   * Custom style for overriding this component's CSS.
   */
  style?: CSSProperties;

  showButton?: boolean;
}
