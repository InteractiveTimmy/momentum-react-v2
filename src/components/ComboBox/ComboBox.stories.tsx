import { MultiTemplate, Template } from '../../storybook/helper.stories.templates';
import { DocumentationPage } from '../../storybook/helper.stories.docs';
import StyleDocs from '../../storybook/docs.stories.style.mdx';
import { Item, Section } from '@react-stately/collections';
import React from 'react';
import ComboBox, { ComboBoxProps } from './';
import argTypes from './ComboBox.stories.args';
import Documentation from './ComboBox.stories.docs.mdx';
import { useListState } from '@react-stately/list';
import ListBoxBase from '../ListBoxBase';
import Flex from '../Flex';

export default {
  title: 'Momentum UI/ComboBox',
  component: ComboBox,
  parameters: {
    expanded: true,
    docs: {
      page: DocumentationPage(Documentation, StyleDocs),
    },
  },
};

const Dual = (props: any) => {
  return (
    <Flex alignItems="center">
      <ComboBox showButton={false}>
        <Item key="1">Red</Item>
        <Item key="2">Blue</Item>
        <Item key="3">Yellow</Item>
      </ComboBox>
      <ComboBox>
        <Item key="4">07377148709</Item>
        <Item key="5">07377142322</Item>
        <Item key="6">07377142322</Item>
      </ComboBox>
    </Flex>
  );
};

// NOTE: Primary story. This renders a single component with all external props.
const Example = Template<any>(Dual).bind({});

Example.argTypes = { ...argTypes };

// TODO: Provide default arguments for this story here. These populate into the argument table for this component.
Example.args = {
  label: 'this',
  children: [<Item key="1">Red</Item>, <Item key="2">Blue</Item>, <Item key="3">Yellow</Item>],
};

const ListBoxExampleComponent = (props: any) => {
  const state = useListState(props);
  return <ListBoxBase state={state} />;
};

// NOTE: Primary story. This renders a single component with all external props.
const ListBoxExample = Template<any>(ListBoxExampleComponent).bind({});

ListBoxExample.argTypes = { ...argTypes };

// TODO: Provide default arguments for this story here. These populate into the argument table for this component.
ListBoxExample.args = {
  label: 'this',
  children: [<Item key="1">Red</Item>, <Item key="2">Blue</Item>, <Item key="3">Yellow</Item>],
};

// TODO: Inject additional stories here.

// NOTE: Common variants story. This renders multiple variants of a single component.
const Common = MultiTemplate<ComboBoxProps>(ComboBox).bind({});

Common.argTypes = { ...argTypes };
delete Common.argTypes.children;

// TODO: Provide default arguments for this story here. These populate into the argument table for this component for all variants.
Common.args = {
  children: 'Example',
};

Common.parameters = {
  variants: [{ children: 'Example A' }, { children: 'Example B' }, { children: 'Example C' }],
};

// NOTE: Export stories here. The first export should be `Example`, and the last export should be `Common`.
export { Example, ListBoxExample, Common };
