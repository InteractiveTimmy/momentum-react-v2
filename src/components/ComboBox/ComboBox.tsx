import React, { ReactElement, useEffect, useState } from 'react';
import classnames from 'classnames';

import { STYLE } from './ComboBox.constants';
import { Props } from './ComboBox.types';
import './ComboBox.style.scss';
import { DismissButton, useOverlay } from '@react-aria/overlays';
import { useSelectState } from '@react-stately/select';

import { useComboBoxState } from '@react-stately/combobox';
import { useComboBox } from '@react-aria/combobox';
import { useFilter } from '@react-aria/i18n';
import ButtonPill from '../ButtonPill';
import Popover from '../Popover';
import ListBoxBase from '../ListBoxBase';
import TextInput from '../TextInput';
import { FocusScope } from 'react-aria';

// eslint-disable-next-line @typescript-eslint/ban-types
function ComboBox<T extends object>(props: Props<T>): ReactElement {
  const { className, id, style, showButton = true } = props;

  // Setup filter function and state.
  const { contains } = useFilter({ sensitivity: 'base' });
  const state = useComboBoxState({ ...props, defaultFilter: contains });

  // Setup refs and get props for child elements.
  const buttonRef = React.useRef(null);
  const inputRef = React.useRef(null);
  const listBoxRef = React.useRef(null);
  const overlayRef = React.useRef<HTMLDivElement>(null);

  const { overlayProps } = useOverlay(
    {
      onClose: () => state.close(),
      shouldCloseOnBlur: true,
      isOpen: state.isOpen,
      isDismissable: true,
    },
    overlayRef
  );

  const { buttonProps, inputProps, listBoxProps } = useComboBox(
    {
      ...props,
      inputRef,
      buttonRef,
      listBoxRef,
      popoverRef: overlayRef,
    },
    state
  );

  const listbox = (
    <FocusScope contain restoreFocus>
      <div ref={overlayRef} {...overlayProps} className="md-select-overlay">
        <ListBoxBase
          state={state}
          {...listBoxProps}
          ref={listBoxRef}
          // eslint-disable-next-line jsx-a11y/no-autofocus
          autoFocus={state.focusStrategy || true}
        />
      </div>
    </FocusScope>
  );

  return (
    <div className={classnames(className, STYLE.wrapper)} id={id} style={style}>
      <input {...inputProps} ref={inputRef} aria-label="shut up" />
      <ButtonPill
        ref={buttonRef}
        outline
        {...buttonProps}
        style={{ display: showButton ? 'initial' : 'none' }}
      >
        ^
      </ButtonPill>
      {state.isOpen && listbox}
    </div>
  );
}

export default ComboBox;
