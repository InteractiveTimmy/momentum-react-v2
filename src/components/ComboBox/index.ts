import { default as ComboBox } from './ComboBox';
import * as CONSTANTS from './ComboBox.constants';
import { Props } from './ComboBox.types';

export { CONSTANTS as COMBO_BOX_CONSTANTS };

export type ComboBoxProps = Props;

export default ComboBox;
