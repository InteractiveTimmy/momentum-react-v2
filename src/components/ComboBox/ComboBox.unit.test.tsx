import React from 'react';
import { mount } from 'enzyme';

import ComboBox, { COMBO_BOX_CONSTANTS as CONSTANTS } from './';

describe('<ComboBox />', () => {
  describe('snapshot', () => {
    it('should match snapshot', () => {
      expect.assertions(1);

      const container = mount(<ComboBox />);

      expect(container).toMatchSnapshot();
    });

    it('should match snapshot with className', () => {
      expect.assertions(1);

      const className = 'example-class';

      const container = mount(<ComboBox className={className} />);

      expect(container).toMatchSnapshot();
    });

    it('should match snapshot with id', () => {
      expect.assertions(1);

      const id = 'example-id';

      const container = mount(<ComboBox id={id} />);

      expect(container).toMatchSnapshot();
    });

    it('should match snapshot with style', () => {
      expect.assertions(1);

      const style = { color: 'pink' };

      const container = mount(<ComboBox style={style} />);

      expect(container).toMatchSnapshot();
    });

    /* ...additional snapshot tests... */
  });

  describe('attributes', () => {
    it('should have its wrapper class', () => {
      expect.assertions(1);

      const element = mount(<ComboBox />)
        .find(ComboBox)
        .getDOMNode();

      expect(element.classList.contains(CONSTANTS.STYLE.wrapper)).toBe(true);
    });

    it('should have provided class when className is provided', () => {
      expect.assertions(1);

      const className = 'example-class';

      const element = mount(<ComboBox className={className} />)
        .find(ComboBox)
        .getDOMNode();

      expect(element.classList.contains(className)).toBe(true);
    });

    it('should have provided id when id is provided', () => {
      expect.assertions(1);

      const id = 'example-id';

      const element = mount(<ComboBox id={id} />)
        .find(ComboBox)
        .getDOMNode();

      expect(element.id).toBe(id);
    });

    it('should have provided style when style is provided', () => {
      expect.assertions(1);

      const style = { color: 'pink' };
      const styleString = 'color: pink;';

      const element = mount(<ComboBox style={style} />)
        .find(ComboBox)
        .getDOMNode();

      expect(element.getAttribute('style')).toBe(styleString);
    });

    /* ...additional attribute tests... */
  });

  describe('actions', () => {
    /* ...action tests... */
  });
});
