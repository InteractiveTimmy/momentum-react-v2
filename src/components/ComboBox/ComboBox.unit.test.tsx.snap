// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ComboBox /> snapshot should match snapshot 1`] = `
<ComboBox>
  <div
    className="md-combo-box-wrapper"
  />
</ComboBox>
`;

exports[`<ComboBox /> snapshot should match snapshot with className 1`] = `
<ComboBox
  className="example-class"
>
  <div
    className="example-class md-combo-box-wrapper"
  />
</ComboBox>
`;

exports[`<ComboBox /> snapshot should match snapshot with id 1`] = `
<ComboBox
  id="example-id"
>
  <div
    className="md-combo-box-wrapper"
    id="example-id"
  />
</ComboBox>
`;

exports[`<ComboBox /> snapshot should match snapshot with style 1`] = `
<ComboBox
  style={
    Object {
      "color": "pink",
    }
  }
>
  <div
    className="md-combo-box-wrapper"
    style={
      Object {
        "color": "pink",
      }
    }
  />
</ComboBox>
`;
